File: mobile-engage/src/main/java/com/emarsys/mobileengage/responsehandler/InAppCleanUpResponseHandler.java
package com.emarsys.mobileengage.responsehandler;

import com.emarsys.core.database.repository.Repository;
import com.emarsys.core.database.repository.SqlSpecification;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClicked;
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIam;
import com.emarsys.mobileengage.iam.model.specification.FilterByCampaignId;
import com.emarsys.mobileengage.util.RequestUrlUtils;

import org.json.JSONArray;
import org.json.JSONObject;

public class InAppCleanUpResponseHandler extends AbstractResponseHandler {

    private static final String OLD_MESSAGES = "old_messages";

    private final Repository<DisplayedIam, SqlSpecification> displayedIamRepository;
    private final Repository<ButtonClicked, SqlSpecification> buttonClickedRepository;

    public InAppCleanUpResponseHandler(
            Repository<DisplayedIam, SqlSpecification> displayedIamRepository,
            Repository<ButtonClicked, SqlSpecification> buttonClickedRepository) {
        Assert.notNull(displayedIamRepository, "DisplayedIamRepository must not be null!");
        Assert.notNull(buttonClickedRepository, "ButtonClickedRepository must not be null!");
        this.displayedIamRepository = displayedIamRepository;
        this.buttonClickedRepository = buttonClickedRepository;
    }

    @Override
    protected boolean shouldHandleResponse(ResponseModel responseModel) {
        boolean shouldHandle = false;

        JSONObject json = responseModel.getParsedBody();
        if (json != null && json.has(OLD_MESSAGES) && isCustomEventResponseModel(responseModel)) {
            JSONArray array = json.optJSONArray(OLD_MESSAGES);
            shouldHandle = array.length() > 0;
        }
        return shouldHandle;
    }

    private boolean isCustomEventResponseModel(ResponseModel responseModel) {
        return RequestUrlUtils.isCustomEvent_V3(responseModel.getRequestModel().getUrl().toString());
    }

    @Override
    protected void handleResponse(ResponseModel responseModel) {
        JSONObject json = responseModel.getParsedBody();
        JSONArray oldMessages = json.optJSONArray(OLD_MESSAGES);
        String[] ids = new String[oldMessages.length()];
        for (int i = 0; i < oldMessages.length(); i++) {
            ids[i] = oldMessages.optString(i);
        }
        displayedIamRepository.remove(new FilterByCampaignId(ids));
        buttonClickedRepository.remove(new FilterByCampaignId(ids));
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/responsehandler/InAppCleanUpResponseHandler.kt
package com.emarsys.mobileengage.responsehandler

import com.emarsys.common.feature.InnerFeature
import com.emarsys.core.database.repository.Repository
import com.emarsys.core.database.repository.SqlSpecification
import com.emarsys.core.feature.FeatureRegistry
import com.emarsys.core.response.AbstractResponseHandler
import com.emarsys.core.response.ResponseModel
import com.emarsys.mobileengage.iam.model.buttonclicked.ButtonClicked
import com.emarsys.mobileengage.iam.model.displayediam.DisplayedIam
import com.emarsys.mobileengage.iam.model.specification.FilterByCampaignId
import com.emarsys.mobileengage.util.RequestModelHelper
import kotlinx.coroutines.runBlocking

class InAppCleanUpResponseHandler(
    private val displayedIamRepository: Repository<DisplayedIam, SqlSpecification>,
    private val buttonClickedRepository: Repository<ButtonClicked, SqlSpecification>,
    private val requestModelHelper: RequestModelHelper
) : AbstractResponseHandler() {

    companion object {
        private const val OLD_MESSAGES = "oldCampaigns"
    }

    override fun shouldHandleResponse(responseModel: ResponseModel): Boolean {
        var shouldHandle = false
        if (!FeatureRegistry.isFeatureEnabled(InnerFeature.EVENT_SERVICE_V4)) {
            val json = responseModel.parsedBody
            if (json != null && json.has(OLD_MESSAGES) && isCustomEventResponseModel(responseModel)) {
                val array = json.optJSONArray(OLD_MESSAGES)
                if (array != null) {
                    shouldHandle = array.length() > 0
                }
            }
        }
        return shouldHandle
    }

    private fun isCustomEventResponseModel(responseModel: ResponseModel): Boolean {
        return requestModelHelper.isCustomEvent(responseModel.requestModel)
    }

    override fun handleResponse(responseModel: ResponseModel) {
        val json = responseModel.parsedBody!!
        val oldMessages = json.optJSONArray(OLD_MESSAGES)
        if (oldMessages != null) {
            val ids = arrayOfNulls<String>(oldMessages.length())
            for (i in 0 until oldMessages.length()) {
                ids[i] = oldMessages.optString(i)
            }
            runBlocking {
                displayedIamRepository.remove(FilterByCampaignId(*ids))
                buttonClickedRepository.remove(FilterByCampaignId(*ids))
            }
        }
    }
}